fix: never call "getResult" more than once per render

aleclarson opened the pull request:
**The problem:** When the `inputs` are undefined, the `getResult` function is called *twice* on the first render.

**The solution:** Pass nothing as the initial value of the `committed` ref, so we can distinguish the first render from every other render. Then assume the `cache` is "current" when `committed.current` is undefined.

aleclarson commented:
**Workaround:** Instead of passing undefined for `inputs`, pass an array with a value that's always different, like `[{}]`.

rimunroe commented:
What's the motivation for this use case? The dependency argument for the base hooks isn't supposed to change type between renders.

```jsx
function Component({ dependencies }) {
  return React.useMemo(() => null, dependencies);
}

render(<Component dependencies={undefined} />);
render(<Component dependencies={[]} />);
```

The above code leads to the following warning:

> Warning: useMemo received a final argument during this render, but not during the previous render. Even though the final argument is optional, its type cannot change between renders.

aleclarson commented on src/index.js:
```diff
@@ -17,16 +18,18 @@ export function useMemoOne<T>(
+    inputs
+      ? committed.current.inputs &&
+          areInputsEqual(inputs, committed.current.inputs)
+      : getResult === initial.getResult,
```
When `inputs` is undefined, this compares `getResult` and `initial.getResult`, which are always equal on the first render. After the first render, they are always *not* equal, unless you wrapped the `getResult` argument with `useCallback` or `useMemo` (which nobody does, right?) or you hoisted the `getResult` function out of the component (so its identity never changes between renders).

If those two edge cases are too sketchy, we could use a nonce instead. On each `useMemoOne` call, the nonce would be incremented, to ensure `initial.nonce` is equal to the current `nonce` on the first render only.

aleclarson commented:
@rimunroe Pretty strange that `React.useMemo` does that, but I guess that means it's not a concern for this PR. Thanks! 👍

rimunroe commented:
Why is it strange? The function which produces the value shouldn't be getting swapped out, so the dependencies should never change. This isn't specific to `useMemo`. Any hook which accepts a dependency array does this. Check out the source code for [`areHookInputsEqual`](https://github.com/facebook/react/blob/v16.9.0/packages/react-reconciler/src/ReactFiberHooks.js#L316-L363).

alexreardon commented:
Sorry for the delay on this. I'll aim to look at this soon

aleclarson commented:
@alexreardon Awesome!

I just reworked this PR to get rid of its previous edge cases (described [here](https://github.com/alexreardon/use-memo-one/pull/10#discussion_r328684859)). It's much cleaner than my previous attempt.

Also, I pushed another commit to this PR, which clears the initial cache once it's replaced (in order to free up memory). Let me know if you prefer to squash PRs, and I'll move that to its own PR. 👍

aleclarson commented:
Bump, in case you forgot about this. ;)

alexreardon commented:
What do you think of this solution? https://github.com/alexreardon/use-memo-one/pull/22

aleclarson commented:
@alexreardon In your solution, there's no way to detect when the initial cache is replaced. See my 2nd commit.

Also, my solution doesn't require an extra useRef.

aleclarson commented:
@alexreardon I'd prefer to avoid the extra useRef, which my solution does. But both solutions get the job done.

This PR also clears the initial cache after it's replaced. See my 2nd commit. Can you cherrypick that over to your PR?

aleclarson commented:
I've separated the test into its own commit, and rewrote the outdated description of the 1st commit. Also, I rearranged the `if` blocks, added braces, and fixed the cache check that I messed up. In addition to avoiding an extra `useRef`, I'm also a fan of how `committed.current` is dereferenced only once.

aleclarson commented:
Ping @alexreardon

alexreardon commented:
Hi @aleclarson, thank you so much for your contribution. I greatly appreciate your effort to build and test a great solution. I have decided to go with #22 which also fixes the issue. I personally prefer having an additional `isFirstRun` `ref` as it makes it super clear to me what is going on. Given I plan on personally maintaining this code I made the tough decision to go with a solution that I personally found more intuitive.

It would be amazing to continue to collaborate with you on this project
